ne/engine.go

```diff
@@ -0,0 +1,229 @@
+package ne
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// NotificationStep is one step of a notification process
+type NotificationStep interface {
+	NotifyMethod() string
+	Frequency() time.Duration
+	Until() time.Duration
+}
+
+// NotificationRequest is an interface consumed by the notification handler. It
+// abstracts the differences between people and teams away.
+type Notification interface {
+	NextStep() (NotificationStep, bool) // returns the next notification step, handling possible escalations, or true if there are not more steps available
+	ID() string                         // return the assigned UUID
+	Message() string                    // return the message (content) to be sent
+	Subject() string                    // return a mesage subject, if availabe, of a configurable default
+}
+
+type Engine struct {
+	Config        Config
+	planChan      chan Notification
+	stopChan      chan string
+	mutex         *sync.Mutex // protects all below
+	stoppers      map[string]chan struct{}
+	notifications map[string]Notification
+	conversations map[string]string
+}
+
+func New(c Config) *Engine {
+	ne := &Engine{
+		Config:        c,
+		planChan:      make(chan Notification, 100),
+		stopChan:      make(chan string, 100),
+		mutex:         &sync.Mutex{},
+		stoppers:      make(map[string]chan struct{}),
+		notifications: make(map[string]Notification),
+		conversations: make(map[string]string),
+	}
+	go ne.start()
+	return ne
+}
+
+func (e *Engine) start() {
+	for {
+		select {
+		case notifyReq := <-e.planChan:
+			e.startNotification(notifyReq)
+		case uuid := <-e.stopChan:
+			e.stopNotification(uuid)
+		}
+	}
+}
+
+func (e *Engine) GetMessage(uuid string) string {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	if nr, found := e.notifications[uuid]; found {
+		return nr.Message()
+	}
+	return ""
+}
+
+func (e *Engine) SetConversation(key, value string) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	e.conversations[key] = value
+}
+
+func (e *Engine) GetConversation(key string) (string, error) {
```

Comment:
Since these are all new methods (to me, at least), would you please add some comments that give a brief description of what they do.   Some things are obvious (EnqueueNotification) but some things are less obvious (SetConversation, for example).